iam-login-service/src/main/java/it/infn/mw/iam/core/web/aup/AupReminderTask.java

```diff
@@ -0,0 +1,110 @@
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import it.infn.mw.iam.notification.NotificationFactory;
+import it.infn.mw.iam.persistence.model.IamAup;
+import it.infn.mw.iam.persistence.model.IamAupSignature;
+import it.infn.mw.iam.persistence.repository.IamAccountRepository;
+import it.infn.mw.iam.persistence.repository.IamAupRepository;
+import it.infn.mw.iam.persistence.repository.IamAupSignatureRepository;
+import it.infn.mw.iam.persistence.repository.IamEmailNotificationRepository;
+
+@Component
+public class AupReminderTask {
+
+  @Autowired
+  IamAccountRepository accounts;
+
+  @Autowired
+  IamAupRepository aupRepo;
+
+  @Autowired
+  NotificationFactory notification;
+
+  @Autowired
+  IamAupSignatureRepository aupSignatureRepo;
+
+  @Autowired
+  IamEmailNotificationRepository emailNotificationRepo;
+
+  public void sendAupReminders() {
+    aupRepo.findDefaultAup().ifPresent(aup -> {
+      LocalDate currentDate = LocalDate.now();
+      LocalDate expirationDate = currentDate.minusDays(aup.getSignatureValidityInDays());
+      Date expirationDateAsDate = toDate(expirationDate);
+      Date expirationDatePlusOneDayAsDate = toDate(expirationDate.plusDays(1));
+      List<Integer> reminderIntervals = parseReminderIntervals(aup.getAupRemindersInDays());
+
+      reminderIntervals.forEach(
+          interval -> processRemindersForInterval(aup, currentDate, interval, expirationDate));
+
+      List<IamAupSignature> expiredSignatures = aupSignatureRepo.findByAupAndSignatureTime(aup,
+          expirationDateAsDate, expirationDatePlusOneDayAsDate);
+
+      // check if an email of type AUP_EXPIRATION does not already exist, because it is never deleted
+      expiredSignatures.forEach(s -> {
+        if (emailNotificationRepo
+          .countAupExpirationMessPerAccount(s.getAccount().getUserInfo().getEmail()) == 0) {
+          notification.createAupSignatureExpMessage(s.getAccount());
+        }
+      });
+    });
+
+  }
+
+  private Date toDate(LocalDate localDate) {
+    return Date.from(localDate.atStartOfDay(ZoneId.systemDefault()).toInstant());
+  }
+
+  private void processRemindersForInterval(IamAup aup, LocalDate currentDate, Integer interval,
+      LocalDate expirationDate) {
+    LocalDate reminderDate = expirationDate.plusDays(interval);
+    Date reminderDateAsDate = toDate(reminderDate);
+    Date reminderDatePlusOneAsDate = toDate(reminderDate.plusDays(1));
+    Date tomorrowAsDate = toDate(currentDate.plusDays(1));
+
+    List<IamAupSignature> signatures = aupSignatureRepo.findByAupAndSignatureTime(aup,
+        reminderDateAsDate, reminderDatePlusOneAsDate);
+
+    // check if an email of type AUP_REMINDER does not already exist, because it is never deleted
+    signatures.forEach(s -> {
+      if (emailNotificationRepo.countAupRemindersPerAccount(s.getAccount().getUserInfo().getEmail(),
```

Comment:
(same as above)

I'd prefer to define a method that hides how this check is implemented but explains with its name the real purpose: for example you could define a method called aupReminderEmailAlreadySentFor(IamAccount a)

This makes code more readable